src/reflect/scala/reflect/internal/Scopes.scala

```diff
@@ -282,6 +282,35 @@ trait Scopes extends api.Scopes { self: SymbolTable =>
       }
     }
 
+    final def lookupSymbolEntry(sym: Symbol): ScopeEntry = {
+      var e = lookupEntry(sym.name)
+      while (e ne null) {
+        if (e.sym == sym) return e
+        e = lookupNextEntry(e)
+      }
+      null
+    }
+
+    final def lookupCompanion(original: Symbol): Symbol = {
+      lookupSymbolEntry(original) match {
+        case null =>
+        case entry if entry.sym == original =>
+          var e = lookupEntry(original.name.companionName)
+          while (e != null) {
+            // 1) Must be owned by the same Scope, to ensure that in
+            //   `{ class C; { ...; object C } }`, the class is not seen as a comaniopn of the object.
+            // 2) Must be in exactly the same scope, so that `{ class C; def C }` are not companions
+            // TODO Exclude type aliases as companions of terms?
```

Comment:
yes, we should. `sym.companionClass` also does since https://github.com/scala/scala/pull/5482/commits